statistics/handle/handle_hist.go

```diff
@@ -0,0 +1,335 @@
+				time.Sleep(time.Millisecond * 10)
+			} else {
+				readerCtx.reader = newReader
+				readerCtx.createdTime = time.Now()
+				return
+			}
+		}
+	} else {
+		return
+	}
+}
+
+// readStatsForOne reads hist for one column, TODO load data via kv-get asynchronously
+func (h *Handle) readStatsForOne(col model.TableColumnID, c *statistics.Column, reader *statsReader) (*statistics.Column, error) {
+	hg, err := h.histogramFromStorage(reader, col.TableID, c.ID, &c.Info.FieldType, c.Histogram.NDV, 0, c.LastUpdateVersion, c.NullCount, c.TotColSize, c.Correlation)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	cms, topN, err := h.cmSketchAndTopNFromStorage(reader, col.TableID, 0, col.ColumnID)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	fms, err := h.fmSketchFromStorage(reader, col.TableID, 0, col.ColumnID)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	rows, _, err := reader.read("select stats_ver from mysql.stats_histograms where is_index = 0 and table_id = %? and hist_id = %?", col.TableID, col.ColumnID)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	if len(rows) == 0 {
+		logutil.BgLogger().Error("fail to get stats version for this histogram", zap.Int64("table_id", col.TableID), zap.Int64("hist_id", col.ColumnID))
+	}
+	colHist := &statistics.Column{
+		PhysicalID: col.TableID,
+		Histogram:  *hg,
+		Info:       c.Info,
+		CMSketch:   cms,
+		TopN:       topN,
+		FMSketch:   fms,
+		IsHandle:   c.IsHandle,
+		StatsVer:   rows[0].GetInt64(0),
+	}
+	// Column.Count is calculated by Column.TotalRowCount(). Hence, we don't set Column.Count when initializing colHist.
+	colHist.Count = int64(colHist.TotalRowCount())
+	return colHist, nil
+}
+
+// drainColTask will hang until a column task can return.
+func (h *Handle) drainColTask(exit chan struct{}) (*NeededColumnTask, error) {
+	// select NeededColumnsCh firstly, if no task, then select TimeoutColumnsCh
+	for {
+		select {
+		case <-exit:
+			return nil, errExit
+		case task, ok := <-h.StatsLoad.NeededColumnsCh:
+			if !ok {
+				return nil, errors.New("drainColTask: cannot read from NeededColumnsCh, maybe the chan is closed.")
+			}
+			if time.Now().After(task.ToTimeout) {
+				h.StatsLoad.TimeoutColumnsCh <- task
+				continue
+			}
+			return task, nil
+		case task, ok := <-h.StatsLoad.TimeoutColumnsCh:
+			select {
+			case <-exit:
+				return nil, errExit
+			case task0, ok0 := <-h.StatsLoad.NeededColumnsCh:
+				if !ok0 {
+					return nil, errors.New("drainColTask: cannot read from NeededColumnsCh, maybe the chan is closed.")
+				}
+				// send task back to TimeoutColumnsCh and return the task drained from NeededColumnsCh
+				h.StatsLoad.TimeoutColumnsCh <- task
+				return task0, nil
+			default:
+				if !ok {
+					return nil, errors.New("drainColTask: cannot read from TimeoutColumnsCh, maybe the chan is closed.")
+				}
+				// NeededColumnsCh is empty now, handle task from TimeoutColumnsCh
+				return task, nil
+			}
```

Comment:
At line 264~267, we put a timeout result to the timeout channel. This will release a position of the needed column channel.
Would it be possible in an extreme situation that there're many timeout tasks and they are all put to the timeout channel?

Then we go to line 278, we want to put a task back to the timeout channel. If the needed column channel put many tasks to the timeout channel. Would it be possible that before we put the task back to the channel, the channel is full again? Then we meet a deadlock...

Reply:
Good catch... WriteTo TimeoutCh should be non-blocking.

Reply:
Good catch... WriteTo TimeoutCh should be non-blocking.